### DISCUSSION TITLE
git error: failed to push some refs to remote

### DISCUSSION THREAD

**Discussion by SuperJUNOed** [community/community#49593] [Actions] [answered]
### Select Topic Area

Question

### Body

I can't push now, though I could do it yesterday.

When I use git push origin master, I get an error:
`
$ git remote -v
origin  https://github.com/REDACTED.git (fetch)
origin  https://github.com/REDACTED.git (push)

$ git push origin master
Username for 'https://github.com': REDACTED
Password for 'https://REDACTED@github.com':
To https://github.com/REDACTED.git
! [rejected]         master -> master (non-fast-forward)
error: failed to push some refs to 'https://github.com/REDACTED.git'
hint: Updates were rejected because the tip of your current branch is behind
hint: its remote counterpart. Integrate the remote changes (e.g.
hint: 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.`

**Comment by jge162** [comment]
Hello @SuperJUNOed . 

- To resolve this issue, you should first run `git pull origin master` to update your local branch with the changes from the remote branch. 

- After resolving any conflicts that may arise during the merge, you should be able to push your changes to the remote branch using `git push origin master`.

**Reply by conceptfac** [reply]
fatal: couldn't find remote ref master

**Reply by ishupaldhar** [reply]
****

**Reply by stevemwangi** [reply]
git push --set-upstream origin master

**Comment by s-spoon** [comment, marked as answer]
And you can [rename existing repository default branch from master to main](https://github.com/github/renaming).
The rest of this 2014 answer has been updated to use "main")

(The following assumes github.com itself is not down, as [eri0o](https://stackoverflow.com/users/965638/eri0o) points out in [the comments](https://stackoverflow.com/questions/24114676/git-error-failed-to-push-some-refs-to-remote/24114760#comment121465079_24114676): see [www.githubstatus.com](https://www.githubstatus.com/) to be sure)

If the GitHub repo has seen new commits pushed to it, while you were working locally, I would advise using:

> git pull --rebase
> git push

The full syntax is:

>git pull --rebase origin main
>git push origin main

[With Git 2.6+](https://stackoverflow.com/a/30209750/6309) (Sept. 2015), after having done (once)

>git config --global pull.rebase true
>git config --global rebase.autoStash true

A simple git pull would be enough.
(Note: with [Git 2.27 Q2 2020](https://stackoverflow.com/a/61562652/6309), a merge.autostash is also available for your regular pull, without rebase)

That way, you would replay (the --rebase part) your local commits on top of the newly updated origin/main (or origin/yourBranch: git pull origin yourBranch).

See a more complete example in the [chapter 6 Pull with rebase](http://chimera.labs.oreilly.com/books/1230000000561/ch06.html#pull-rebase) of the [Git Pocket Book](http://chimera.labs.oreilly.com/books/1230000000561).

I would recommend a:

># add and commit first
>#
>git push -u origin main

># Or git 2.37 Q2 2022+
>git config --global push.autoSetupRemote true
>git push

That would establish a tracking relationship between your local main branch and its upstream branch.
After that, any future push for that branch can be done with a simple:

>git push

Again, with [Git 2.37+ and its global option push.autoSetupRemote](https://stackoverflow.com/a/72401899/6309), a simple git push even for the first one would do the same (I.e: establishing a tracking relationship between your local main branch and its upstream branch origin/main).

Since the OP already [reset and redone its commit](https://stackoverflow.com/a/18589043/6309) on top of origin/main:

>git reset --mixed origin/main
>git add .
>git commit -m "This is a new commit for what I originally planned to be amended"
>git push origin main

There is no need to pull --rebase.

Note: git reset --mixed origin/main can also be written git reset origin/main, since the --mixed option is the default one when using [git reset](http://git-scm.com/docs/git-reset).

**Reply by ghost** [reply]
thank you sir

**Reply by bhawna007** [reply]
Thank you so much

**Reply by abiolafolalu** [reply]
Thank you so much, sir

**Reply by abhinavtyagi88** [reply]
thanks a lot sir

**Reply by Artternative** [reply]
Thank you so much.

**Reply by ghost** [reply]
Nada disso resolveu meu problema

**Reply by smoin55** [reply]
Yeah its working Thanks !!!!

**Comment by RobQuistNL** [comment]
It seems like right now there might be some issues on Github that cause these issues. I've seen errors regarding pushing, but then the commit actually did make it in the repository.

The actions also seem to be broken - I see no status or information about them.

https://www.githubstatus.com/ shows some issues;
![image](https://github.com/community/community/assets/1442796/20f2bdb8-2880-40a0-aa49-07720ce253d5)


https://downdetector.com/status/github/
![image](https://github.com/community/community/assets/1442796/2e569e27-0cfb-4836-bcfe-182cd9f7035f)

**Reply by tochman** [reply]
Great. So it's not just me 🥲

**Reply by almahdi404** [reply]
In my case commits are being pushed to github.
git clone does include the pushed commits but they aren't visible in the repository.

**Reply by ruk93** [reply]
Having same experience as @almahdi404 mentioned.

**Reply by brenicolas** [reply]
experiencing the same issue @almahdi404

**Comment by ntruc710i** [comment]
git pull --rebase origin main

git push -u origin main

**Reply by RajeshMoravaneni7** [reply]
It did work for me. Thank You

**Reply by Jesusson-softwareandIT** [reply]
Wow guy you put the p in programmer. Nice one. Works

**Reply by suchir-cyber** [reply]
By trying this , it really worked for me . Thanks

**Comment by Sudhanshu-273** [comment]
I got the same error, The reason behind it was, I exposed one of my API keys in a working file, I moved it to env, and it worked for me.
